d2core/d2checkbox/checkbox.go

```diff
@@ -0,0 +1,155 @@
+	SpritePath       string
+	PalettePath      string
+	FontPath         string
+	ClickableRect    *rectangle.Rectangle
+	XSegments        int
+	YSegments        int
+	BaseFrame        int
+	DisabledFrame    int
+	DisabledColor    uint32
+	TextOffset       float64
+	FixedWidth       int
+	FixedHeight      int
+	LabelColor       uint32
+	Toggleable       bool
+	AllowFrameChange bool
+	HasImage         bool
+	Tooltip          int
+	TooltipXOffset   int
+	TooltipYOffset   int
+}
+
+// New creates an instance of Button
+func New() *Checkbox {
+	checkbox := &Checkbox{
+		Layout: GetDefaultLayout(),
+	}
+
+	return checkbox
+}
+
+// GetDefaultLayout returns the default layout of a checkbox.
+func GetDefaultLayout() CheckboxLayout {
+	return CheckboxLayout{
+		X:                0,
+		Y:                0,
+		SpritePath:       d2resource.Checkbox,
+		PalettePath:      d2resource.PaletteFechar,
+		FontPath:         d2resource.FontExocet10,
+		XSegments:        1,
+		YSegments:        1,
+		BaseFrame:        0,
+		DisabledFrame:    -1,
+		DisabledColor:    lightGreyAlpha75,
+		TextOffset:       CheckboxDefaultTextOffset,
+		FixedWidth:       CheckboxDefaultWidth,
+		FixedHeight:      CheckboxDefaultHeight,
+		LabelColor:       goldAlpha100,
+		Toggleable:       true,
+		AllowFrameChange: true,
+		HasImage:         true,
+		Tooltip:          0,
+		TooltipXOffset:   0,
+		TooltipYOffset:   0,
+	}
+}
+
+// OnActivated defines the callback handler for the activate event
+func (v *Checkbox) OnActivated(callback callbackFunc) {
+	v.callback = callback
+}
+
+// Activate calls the on activated callback handler, if any
+func (v *Checkbox) Activate(thisComponent akara.Component) bool {
+	if v.GetEnabled() {
+		v.Toggle()
+	}
+
+	if v.callback != nil {
+		return v.callback(thisComponent)
+	}
+
+	return false
+}
+
+// Toggle negates the toggled state of the button
+func (v *Checkbox) Toggle() {
+	v.SetPressed(!v.GetPressed())
+}
+
+// GetEnabled returns the enabled state
+func (v *Checkbox) GetEnabled() bool {
+	return v.enabled
+}
+
+// SetEnabled sets the enabled state
+func (v *Checkbox) SetEnabled(enabled bool) {
+	v.enabled = enabled
+}
+
+// GetPressed returns the enabled state
+func (v *Checkbox) GetPressed() bool {
+	return v.pressed
+}
+
+// SetPressed sets the enabled state
+func (v *Checkbox) SetPressed(pressed bool) {
+	v.pressed = pressed
+}
+
+// Update updates the checkbox's sprite in accordance with the checkbox's current state.
+// This ensures that the checkbox rendered in the UI accurately reflects the state of the checkbox.
+func (v *Checkbox) Update() {
+	if v.Sprite == nil {
+		return
+	}
+
+	switch {
+	case v.GetEnabled() && v.GetPressed():
+		// checked, enabled
+		_ = v.Sprite.SetCurrentFrame(1)
+	case v.GetEnabled():
+		// unchecked, enabled
+		_ = v.Sprite.SetCurrentFrame(0)
+	case v.GetPressed():
+		// checked, disabled
+		_ = v.Sprite.SetCurrentFrame(1)
+	default:
+		// unchecked, disabled
+		_ = v.Sprite.SetCurrentFrame(0)
+	}
```

Comment:
it looks like this switch can be boiled down a bit more, but maybe i'm not looking at it correctly... or maybe it just makes a lot more sense to keep all of the cases separate for readability.

Reply:
This is where I was messing with the setting the sprite colors and effects, to properly render a "disabled" checkbox. I couldn't get anything working, so I ended up stripping it out, but my thought was that if the checkbox is disabled, the color should be changed to silver or something, like it is in retail D2

Reply:
I personally prefer to keep the cases separate in case this eventually gets added in the future (and it's less work for me!)

Reply:
that's a good reason to keep these cases separate, i don't have any issues with that.